File: submissions/ccc22j5.cpp
#include <bits/stdc++.h>

// 2/14/2023
using namespace std;
typedef pair<int,int> pii;
int main() {
    cin.sync_with_stdio(0);
    cin.tie(0);
    int n,t;cin>>n>>t;
    pii trees[t+2];
    for (int i = 0; i < t; ++i) {
        int r,c;cin>>r>>c;
        trees[i]=pii(r,c);
    }
    trees[t]=pii(0,0);
    trees[t+1]=pii(n+1, n+1);
    int ans = 0;
    for (auto& left : trees) {
        for (auto& right : trees) {
            if (left==right||left.second>right.second) {
                continue;
            }
            vector<pii> inside;
            for (auto& tree : trees) {
                if (tree.second > left.second&&tree.second<right.second) {
                    inside.push_back(tree);
                }
            }
            inside.push_back(pii(0,0));
            inside.push_back(pii(n+1, n+1));
            sort(inside.begin(), inside.end());
            for (int i = 0; i < inside.size()-1; i++) {
                ans = max(ans, min(abs(left.second-right.second)-1,abs(inside[i+1].first-inside[i].first)-1));
            }
        }
    }
    cout << ans << "\n";
}

File: submissions/ccc22j5.java
import java.io.*;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Objects;
import java.util.StringTokenizer;

public class SquarePool {
    private static final BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
    private static final PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
    private static StringTokenizer stringTokenizer;

    private static class Coordinate {
        private int row;
        private int column;

        private Coordinate(int row, int column) {
            this.row = row;
            this.column = column;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Coordinate that = (Coordinate) o;
            return row == that.row && column == that.column;
        }

        @Override
        public int hashCode() {
            return Objects.hash(row, column);
        }

        @Override
        public String toString() {
            return row + " " + column;
        }
    }

    public static void main(String[] args) throws IOException {
        int length = readInt();
        int numberOfTrees = readInt();
        Coordinate[] trees = new Coordinate[numberOfTrees + 2];
        for (int i = 0; i < numberOfTrees; i++) {
            int row = readInt(), column = readInt();
            trees[i] = new Coordinate(row, column);
        }
        trees[numberOfTrees] = new Coordinate(0, 0);
        trees[numberOfTrees + 1] = new Coordinate(length + 1, length + 1);
        int answer = 0;
        for (Coordinate left : trees) {
            for (Coordinate right : trees) {
                if (left == right || left.column > right.column) {
                    continue;
                }
                ArrayList<Coordinate> insideTrees = new ArrayList<>();
                for (Coordinate tree : trees) {
                    if (tree.column < right.column && tree.column > left.column) {
                        insideTrees.add(tree);
                    }
                }
                insideTrees.add(new Coordinate(0, 0));
                insideTrees.add(new Coordinate(length + 1, length + 1));
                int sideLength = Math.abs(left.column - right.column) - 1;

                Comparator<Coordinate> rowComparator = Comparator.comparingInt(o -> o.row);
                insideTrees.sort(rowComparator);
                for (int i = 0; i < insideTrees.size() - 1; i++) {
                    answer = Math.max(answer, Math.min(sideLength, Math.abs(insideTrees.get(i).row - insideTrees.get(i + 1).row) - 1));
                }
            }
        }

        pr.println(answer);
        pr.close();

    }

    private static String next() throws IOException {
        while (stringTokenizer == null || !stringTokenizer.hasMoreTokens())
            stringTokenizer = new StringTokenizer(bufferedReader.readLine().trim());
        return stringTokenizer.nextToken();
    }

    @SuppressWarnings("unused")
    private static long readLong() throws IOException {
        return Long.parseLong(next());
    }

    @SuppressWarnings("unused")
    private static int readInt() throws IOException {
        return Integer.parseInt(next());
    }

    @SuppressWarnings("unused")
    private static double readDouble() throws IOException {
        return Double.parseDouble(next());
    }

    /*
    @SuppressWarnings("unused")
    private static char readCharacter() throws IOException {
        return next().charAt(0);
    }
    */

    @SuppressWarnings("unused")
    private static String readLine() throws IOException {
        return bufferedReader.readLine().trim();
    }
}